### PR TITLE
Is lc equiv div fix

### PR DESCRIPTION
**Pull request by DaveDRoberts**
Includes previous Python 3 compatibility fixes and fixes division in is_lc_equiv to ensure that an integer is returned.

### CODE REVIEW

File: gsc/explore_lc_orbit.py [hunk 1]
```diff
@@ -145,10 +145,10 @@ def queued_orbit_search(init_graph, local_ops, save_edges, verbose):
         visited += 1
         # Gets next graph on queue and finds representative nodes
         graph_label = queue.pop()
-        graph = class_graph.node[graph_label]['nx_graph']
```
Comment (sammorley-short) [inline gsc/explore_lc_orbit.py:148]: It seems the deprecation of `node` in favour of `nodes` is a NetworkX 2.0 feature, so we should peg the version in `setup.py` to `networkx>=2.0.0`.

File: gsc/explore_lc_orbit.py [hunk 2]
```diff
@@ -145,10 +145,10 @@ def queued_orbit_search(init_graph, local_ops, save_edges, verbose):
...
+        graph = class_graph.nodes[graph_label]['nx_graph']
         node_equivs = find_rep_nodes(graph)
         # Applies set of local ops to each representative node
-        for rep_node, equiv_nodes in node_equivs.iteritems():
```
Comment (sammorley-short) [inline gsc/explore_lc_orbit.py:151]: What was the reasoning for replacing the generator over key-value items with the explicit generation? For speed purposes it's usually better to keep things as generators unless there's a reason for generating the whole set at once.

Comment (sammorley-short) [reply gsc/explore_lc_orbit.py:151]: Ah, I see that iteritems is now deprecated, nevermind!

File: gsc/get_nauty.py [hunk 3]
```diff
@@ -13,9 +13,9 @@ def qudit_graph_map(nx_wg, partition=None):
     For prime-power graph states, can colour by member or family.
     """
     # Gets list of all nodes by layer
-    us, vs, weights = zip(*nx_wg.edges.data('weight'))
+    us, vs, weights = list(zip(*nx_wg.edges.data('weight')))
```
Comment (sammorley-short) [inline gsc/get_nauty.py]: Why do we need to generate the whole list here? Same goes for all the other uses of `list` in this module

File: gsc/is_lc_equiv.py [hunk 4]
```diff
@@ -135,3 +135,4 @@ def are_lc_equiv(g1, g2):
         return True, V
     else:
         return False, None
+
```
Comment (sammorley-short) [inline gsc/is_lc_equiv.py]: Nit: extra line here.
```suggestion
```

**Review by sammorley-short — CHANGES_REQUESTED** [review]
Mostly looks good to me. Just a question over whether we need to be changing some of the generators into lists.

File: gsc/get_nauty.py [hunk 5]
```diff
@@ -101,7 +101,7 @@ def hash_graph(graph):
 
 def canonical_relabel(nx_g):
     """ Returns isomorphic graph with canonical relabelling """
-    nodes, neighs = zip(*nx_g.adjacency())
+    nodes, neighs = list(zip(*nx_g.adjacency()))
```
Comment (DaveDRoberts) [inline gsc/get_nauty.py]: Don't need this change.

File: gsc/get_nauty.py (hunk 3 above)
Comment (DaveDRoberts) [reply gsc/get_nauty.py]: Recasting the zip object as a list isn't necessary here but in some instances it is. I just included it here to be safe, but there's no need to include it here. I'll add comments to all the necessary ones.

File: gsc/get_nauty.py [hunk 6]
```diff
@@ -13,9 +13,9 @@ def qudit_graph_map(nx_wg, partition=None):
     For prime-power graph states, can colour by member or family.
     """
     # Gets list of all nodes by layer
-    us, vs, weights = zip(*nx_wg.edges.data('weight'))
+    us, vs, weights = list(zip(*nx_wg.edges.data('weight')))
     n_layers = int(log(max(weights), 2)) + 1
-    layers = range(n_layers)
+    layers = list(range(n_layers))
```
Comment (DaveDRoberts) [inline gsc/get_nauty.py]: Not required.

File: gsc/get_nauty.py [hunk 7]
```diff
@@ -69,11 +69,11 @@ def convert_nx_to_pyn(nx_g, partition=None):
     if nx_g.__dict__.get('dimension', 2) > 2:
         nx_g, coloring = qudit_graph_map(nx_g, partition)
     # Relabels nodes with integers for compatibility with Pynauty
-    nodes, neighs = zip(*nx_g.adjacency())
+    nodes, neighs = list(zip(*nx_g.adjacency()))
     to_int_node_map = {n: i for i, n in enumerate(nodes)}
     relabel = to_int_node_map.get
-    nodes = map(relabel, nodes)
-    neighs = [map(relabel, node_neighs.keys()) for node_neighs in neighs]
+    nodes = list(map(relabel, nodes))
```
Comment (DaveDRoberts) [inline gsc/get_nauty.py]: nodes does not need to be recast as list type

File: gsc/get_nauty.py [hunk 8]
```diff
@@ -69,11 +69,11 @@ def convert_nx_to_pyn(nx_g, partition=None):
...
+    neighs = [list(map(relabel, node_neighs.keys())) for node_neighs in neighs]
```
Comment (DaveDRoberts) [inline gsc/get_nauty.py:76]: list is necessary here.

File: gsc/is_lc_equiv.py [hunk 9]
```diff
@@ -76,8 +76,8 @@ def GF2nullspace(A):
     Id = np.eye(n, dtype=int)
     for i in range(n):
         while A[:, i].tolist() != Id[:, i].tolist():
-            perm = range(i, m)
-            A[:, perm] = A[:, range(i + 1, m) + [i]]
+            perm = list(range(i, m))
```
Comment (DaveDRoberts) [inline gsc/is_lc_equiv.py:79]: Needed because only lists can be concatenated with lists

File: gsc/is_lc_equiv.py [hunk 10]
```diff
@@ -76,8 +76,8 @@ def GF2nullspace(A):
...
+            A[:, perm] = A[:, list(range(i + 1, m)) + [i]]
```
Comment (DaveDRoberts) [inline gsc/is_lc_equiv.py:80]: Needed because only lists can be concatenated with lists

File: gsc/is_lc_equiv.py [hunk 11]
```diff
@@ -109,9 +109,9 @@ def are_lc_equiv(g1, g2):
     C = sp.symbols('c:' + str(dim1), bool=True)
     D = sp.symbols('d:' + str(dim1), bool=True)
     # Defines solution matrix basis
-    abcd = flatten(zip(A, B, C, D))
+    abcd = flatten(list(zip(A, B, C, D)))
```
Comment (DaveDRoberts) [inline gsc/is_lc_equiv.py]: List() not required.

File: gsc/utils.py [hunk 12]
```diff
@@ -66,7 +66,7 @@ def vector_scale(u, s):
 
 def vector_add(u, v):
     """ Adds vectors u and v """
-    return map(sum, zip(u, v))
+    return list(map(sum, zip(u, v)))
```
Comment (DaveDRoberts) [inline gsc/utils.py]: Don't need list here

File: gsc/graph_builders.py [hunk 13]
```diff
@@ -50,7 +50,7 @@ def make_crazy(graph, n):
                            for u, v in graph.edges()])
     # Creates graph and adds edges and nodes
     crazy_graph = nx.Graph()
-    crazy_graph.add_nodes_from(flatten(crazy_nodes.values()))
+    crazy_graph.add_nodes_from(flatten(list(crazy_nodes.values())))
```
Comment (DaveDRoberts) [inline gsc/graph_builders.py]: list() not need here.

File: gsc/get_nauty.py (hunk 3 above)
Comment (sammorley-short) [reply gsc/get_nauty.py]: Great, thanks a bunch!

**Comment by sammorley-short** [comment]
Thanks for the comments on the `list` casting Dave. So if you can push an updated version with those changes and the others above then this looks good to go :)

**Review by DaveDRoberts — COMMENTED** [review]
Removed unnecessary list recasts from original pull request

**Comment by ivelieu** [comment]
Just going to point out, this pull request still fails some of the tests 

```
short test summary info

FAILED tests/test_README_examples.py::test_LC_explore_example - AssertionError: assert {'links': [{'...id': 5}, ...]} == {'links': [{'...id': 5}, ...]}
FAILED tests/test_README_examples.py::test_prime_dimension_explore_example - TypeError: '>' not supported between instances of 'DegreeView' and 'int'
FAILED tests/test_explore_class.py::test_ququart_pair - TypeError: '>' not supported between instances of 'DegreeView' and 'int'
FAILED tests/test_get_nauty.py::test_find_rep_nodes - AttributeError: 'dict' object has no attribute 'iteritems'
FAILED tests/test_get_nauty.py::test_canonical_relabel - assert ((0, 3), (1, ...1, 3), (2, 3)) == ((0, 1), (0, ...0, 3), (1, 2))
FAILED tests/test_get_nauty.py::test_prime_power_hash_examples - assert 217157435438440218 == 2941527429935884599

6 failed, 12 passed, 2 skipped in 12.65s 
```

test_canonical_relabel, test_LC_explore_example failing corresponds to #2 
test_prime_power_hash_examples failing corresponds to #4 

Looks like the TypeErrors and AttributeError still need fixing though.

File: gsc/is_lc_equiv.py (hunk 11 above)
Comment (sammorley-short) [reply gsc/is_lc_equiv.py]: Cool, please remove then :)

File: gsc/utils.py (hunk 12 above)
Comment (sammorley-short) [reply gsc/utils.py]: Cheers. Please remove then

**Review by sammorley-short — APPROVED** [review]

**Comment by sammorley-short** [comment]
@DaveDRoberts: Thanks for your work here!
@ivelieu: Thanks for the reminder---I had totally forgot about these and will try to fix them soonish. But before I get to these I'd like to have GitHub actions set up, so we can have a reproducible test environment for these jobs. Am working on this here: https://github.com/sammorley-short/gsc/pull/8. In the meantime, if you think you'd be able to take a shot at fixing them, please do! 🙂

**Comment by sammorley-short** [comment]
@DaveDRoberts @ivelieu: FYI repo now has tests run on push, which you can find the results of here: https://github.com/sammorley-short/gsc/actions. Currently, these are failing due to the aforementioned bugs, but at least now we have a single source of truth to test against.

Not sure when I'll have the time to get to fixing the bugs, so if you need them addressed soon then you might want to have a go yourself. LMK if you do and I am happy to help along the way if so.

**Pull request merged**